src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs

```diff
@@ -12688,5 +12942,640 @@ public void RefStructHandler_DynamicInHole(string expression)
+            vbComp.VerifyDiagnostics();
+
+            var code = @"
+M m = null;
+
+m($"""");
+";
+
+            var comp = CreateCompilation(code, references: new[] { vbComp.EmitToImageReference() });
+            comp.VerifyDiagnostics(
+                // (4,3): error CS8949: The InterpolatedStringHandlerArgumentAttribute applied to parameter 'CustomHandler' is malformed and cannot be interpreted. Construct an instance of 'CustomHandler' manually.
+                // m($"");
+                Diagnostic(ErrorCode.ERR_InterpolatedStringHandlerArgumentAttributeMalformed, @"$""""").WithArguments("CustomHandler", "CustomHandler").WithLocation(4, 3),
+                // (4,3): error CS1729: 'CustomHandler' does not contain a constructor that takes 2 arguments
+                // m($"");
+                Diagnostic(ErrorCode.ERR_BadCtorArgCount, @"$""""").WithArguments("CustomHandler", "2").WithLocation(4, 3),
+                // (4,3): error CS1729: 'CustomHandler' does not contain a constructor that takes 3 arguments
+                // m($"");
+                Diagnostic(ErrorCode.ERR_BadCtorArgCount, @"$""""").WithArguments("CustomHandler", "3").WithLocation(4, 3)
+            );
+        }
+
+        [Fact]
+        public void ArgumentsOnDelegateTypes_03()
+        {
+            var code = @"
+using System;
+using System.Runtime.CompilerServices;
+
+M m = (i, c) => Console.WriteLine(c.ToString());
+
+m(1, $"""");
+
+delegate void M(int i, [InterpolatedStringHandlerArgument(""i"")] CustomHandler c);
+
+partial struct CustomHandler
+{
+    public CustomHandler(int literalLength, int formattedCount, int i) : this(literalLength, formattedCount) => _builder.Append(""i:"" + i.ToString());
+}
+";
+
+            var verifier = CompileAndVerify(new[] { code, InterpolatedStringHandlerArgumentAttribute, GetInterpolatedStringCustomHandlerType("CustomHandler", "partial struct", useBoolReturns: false) }, expectedOutput: @"i:1");
+            verifier.VerifyDiagnostics();
+
+            verifier.VerifyIL("<top-level-statements-entry-point>", @"
+{
+  // Code size       48 (0x30)
+  .maxstack  5
+  .locals init (int V_0)
+  IL_0000:  ldsfld     ""M Program.<>c.<>9__0_0""
+  IL_0005:  dup
+  IL_0006:  brtrue.s   IL_001f
+  IL_0008:  pop
+  IL_0009:  ldsfld     ""Program.<>c Program.<>c.<>9""
+  IL_000e:  ldftn      ""void Program.<>c.<<Main>$>b__0_0(int, CustomHandler)""
+  IL_0014:  newobj     ""M..ctor(object, System.IntPtr)""
+  IL_0019:  dup
+  IL_001a:  stsfld     ""M Program.<>c.<>9__0_0""
+  IL_001f:  ldc.i4.1
+  IL_0020:  stloc.0
+  IL_0021:  ldloc.0
+  IL_0022:  ldc.i4.0
+  IL_0023:  ldc.i4.0
+  IL_0024:  ldloc.0
+  IL_0025:  newobj     ""CustomHandler..ctor(int, int, int)""
+  IL_002a:  callvirt   ""void M.Invoke(int, CustomHandler)""
+  IL_002f:  ret
+}
+");
+        }
+
+        [Fact]
+        public void HandlerConstructorWithDefaultArgument_01()
```

Comment:
nit: consider testing constructor with params `(int, int, params object[])` (probably works) or `(int, params int[])` (error?)

Reply:
I'm not entirely sure what you're trying to exercise. None of those types are interpolated string handlers. Can you please show a specific test case?

Reply:
Spelling out what I mean:
```
// I assume this will work
[InterpolatedStringHandler]
partial struct CustomHandler
{
    public CustomHandler(int literalLength, int formattedCount, params object[]) => ...;
}
```

```
// I assume this will error
[InterpolatedStringHandler]
partial struct CustomHandler
{
    public CustomHandler(int literalLength, params int[] formattedCount) => ...;
}
```

Reply:
Ah, now I see. Thanks for the clarification.

Reply:
> ```
> // I assume this will work
> [InterpolatedStringHandler]
> partial struct CustomHandler
> {
>     public CustomHandler(int literalLength, int formattedCount, params object[]) => ...;
> }
> ```

Correct.


> ```
> // I assume this will error
> [InterpolatedStringHandler]
> partial struct CustomHandler
> {
>     public CustomHandler(int literalLength, params int[] formattedCount) => ...;
> }
> ```

Nope, that works fine. We just look for applicable constructors. That constructor is applicable.

Reply:
See https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/improved-interpolated-strings.md#constructor-resolution: specifically, we don't name parameters, and we don't change the constructor resolution rules. We just look for a constructor that can accept two int parameters and a bool. When that fails, we look for a constructor that can accept two int parameters. We say nothing about whether it needs to be normal form or not.